kvdb/cache.go

```diff
@@ -0,0 +1,930 @@
+package kvdb
+
+import (
+	"container/list"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+	"github.com/google/btree"
+)
+
+const (
+	treeDeg = 3
+)
+
+type cacheBucket struct {
+	seq  *uint64
+	tree *btree.BTree
+}
+
+func newCacheBucket() *cacheBucket {
+	return &cacheBucket{
+		tree: btree.New(treeDeg),
+	}
+}
+
+func (c *cacheBucket) get(key []byte) *cachedItem {
+	if key == nil {
+		return nil
+	}
+
+	keyItem := &cachedItem{
+		key: string(key),
+	}
+
+	if valItem := c.tree.Get(keyItem); valItem != nil {
+		return valItem.(*cachedItem)
+	}
+
+	return nil
+}
+
+func (c *cacheBucket) put(item *cachedItem) {
+	c.tree.ReplaceOrInsert(item)
+}
+
+func (c *cacheBucket) del(item *cachedItem) {
+	c.tree.Delete(item)
+}
+
+type cachedItem struct {
+	key    string
+	value  string
+	bucket *cacheBucket
+}
+
+func (c *cachedItem) Less(than btree.Item) bool {
+	return c.key < than.(*cachedItem).key
+}
+
+// Enforce that Cache implements the ExtendedBackend interface.
+var _ walletdb.DB = (*Cache)(nil)
+
+type pendingChange interface {
+	// Reverts the pending change.
+	Revert()
+}
+
+type pendingAdd struct {
+	parent   *cachedItem
+	newChild *cachedItem
+}
+
+func (p *pendingAdd) Revert() {
+	p.parent.bucket.tree.Delete(p.newChild)
+}
+
+type pendingUpdate struct {
+	parent   *cachedItem
+	oldValue string
+}
+
+func (p *pendingUpdate) Revert() {
+	p.parent.value = p.oldValue
+}
+
+type pendingDelete struct {
+	parent   *cachedItem
+	oldChild *cachedItem
+}
+
+func (p *pendingDelete) Revert() {
+	p.parent.bucket.tree.ReplaceOrInsert(p.oldChild)
+}
+
+// Cache is a simple write through cache implementing the kvdb.Backend
+// interface. It's capable of recursively caching top-level buckets speeding up
+// reads by reducing roundtrips to the actual DB backend while remaining
+// consistend with DB state as long as the cache is used when mutating those
+// buckets.
```

Comment:
Might be useful to have some "admin" commands to partially/fully invalidate the cache. Thinking about the situation where a user needs to manually edit the DB to fix some issue (given the DB is replicated they can modify it while lnd has it "open"). Likely not something we need now, but shouldn't be too difficult to implement in the future.

Reply:
Indeed, I haven't thought about this use case. Will add.